### PR TITLE
Add config docs generate command

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
Add a subcommand, so that other components can use this to generate config docs. 
For example:
```
flyteadmin config docs
flytepropeller config docs
```

https://flyte--1608.org.readthedocs.build/en/1608/deployment/cluster_config/flytepropeller_config.html#flytepropeller-config-specification
## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Output table for each config struct in the root section
```
+-------------+------------+--------------------------------+
|    NAME     |    TYPE    |          DESCRIPTION           |
+-------------+------------+--------------------------------+
| endpoint    | config.URL | URL for storage client to      |
|             |            | connect to.                    |
+-------------+------------+--------------------------------+
| auth-type   | string     | Auth Type to use               |
|             |            | [iam,accesskey].               |
+-------------+------------+--------------------------------+
| access-key  | string     | Access key to use. Only        |
|             |            | required when authtype is set  |
|             |            | to accesskey.                  |
+-------------+------------+--------------------------------+
| secret-key  | string     | Secret to use when accesskey   |
|             |            | is set.                        |
+-------------+------------+--------------------------------+
| region      | string     | Region to connect to.          |
+-------------+------------+--------------------------------+
| disable-ssl | bool       | Disables SSL connection.       |
|             |            | Should only be used for        |
|             |            | development.                   |
+-------------+------------+--------------------------------+
```

## Tracking Issue
Related to https://github.com/flyteorg/flyte/pull/1608

## Follow-up issue
_NA_

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -378,6 +378,7 @@ github.com/mattn/go-isatty v0.0.3/go.mod h1:M+lRXTBqGeGNdLjl/ufCoiOlB5xdOkqRJdNx
 github.com/mattn/go-isatty v0.0.4/go.mod h1:M+lRXTBqGeGNdLjl/ufCoiOlB5xdOkqRJdNxMWT7Zi4=
 github.com/mattn/go-isatty v0.0.12 h1:wuysRhFDzyxgEmMf5xjvJ2M9dZoWAXNNr5LSBS7uHXY=
 github.com/mattn/go-isatty v0.0.12/go.mod h1:cbi8OIDigv2wuxKPP5vlRcQ1OAZbq2CE4Kysco4FUpU=
+github.com/mattn/go-runewidth v0.0.2 h1:UnlwIPBGaTZfPQ6T1IGzPI0EkYAQmT9fAEJ/poFC63o=
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply go.sum:381]: ran it, but we still have this dependency.

File: config/config_cmd.go [hunk 2]
```diff
@@ -75,6 +103,92 @@ func redirectStdOut() (old, new *os.File) {
...
 
+func printConfigTable(b interface{}, sectionName string, subsection bool, sectionKeyMap map[string]bool) {
+	val := reflect.Indirect(reflect.ValueOf(b))
+
+	if val.Kind() == reflect.Slice {
+		val = reflect.Indirect(reflect.ValueOf(val.Index(0).Interface()))
+	}
+
+	table := tablewriter.NewWriter(os.Stdout)
+	table.SetHeader([]string{"Name", "Type", "Description"})
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply config/config_cmd.go]: added it.

File: config/config_cmd.go [hunk 3]
```diff
@@ -75,6 +103,92 @@ func redirectStdOut() (old, new *os.File) {
...
+			tagType = t.Type.Elem()
+		}
+
+		fieldName := t.Name
+		fieldType := tagType.String()
+		fieldDescription := ""
+
+		if t.Type.Kind() == reflect.Struct {
+			ss := strings.Split(fieldType, ".")
+			fieldType = ss[len(ss)-1]
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply config/config_cmd.go]: added an if statement to check it

File: config/config_cmd.go [hunk 4]
```diff
@@ -75,6 +103,92 @@ func redirectStdOut() (old, new *os.File) {
...
+		}
+
+		if jsonTag := t.Tag.Get("json"); jsonTag != "" && jsonTag != "-" {
+			var commaIdx int
+			if commaIdx = strings.Index(jsonTag, ","); commaIdx < 0 {
+				commaIdx = len(jsonTag)
+			}
+			fieldName = jsonTag[:commaIdx]
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply config/config_cmd.go]: done

File: config/config_cmd.go [hunk 5]
```diff
@@ -75,6 +103,92 @@ func redirectStdOut() (old, new *os.File) {
...
+		}
+
+		if pFlag := t.Tag.Get("pflag"); pFlag != "" && pFlag != "-" {
+			var commaIdx int
+			if commaIdx = strings.Index(pFlag, ","); commaIdx < 0 {
+				commaIdx = -1
+			}
+			fieldDescription = pFlag[commaIdx+1:]
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply config/config_cmd.go]: done

File: config/config_cmd.go [hunk 6]
```diff
@@ -75,6 +103,92 @@ func redirectStdOut() (old, new *os.File) {
...
+			// Make this string as a link
+			fieldType = fmt.Sprintf("%s_", fieldType)
+		}
+
+		data := []string{fieldName, fieldType, fieldDescription}
+		table.Append(data)
+	}
+
+	// Print out the config table
+	fmt.Println(sectionName)
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply config/config_cmd.go]: re:
1. added it
2. There is a map to record the type that have been printed out, so that we will not get duplicate table
3. Fixed it. https://flyte--1608.org.readthedocs.build/en/1608/deployment/cluster_config/flytepropeller_config.html#section-plugins

File: config/config_cmd.go [hunk 7]
```diff
@@ -75,6 +103,92 @@ func redirectStdOut() (old, new *os.File) {
...
+	table.SetRowLine(true)
+
+	for i := 0; i < val.Type().NumField(); i++ {
+		t := val.Type().Field(i)
+		tagType := t.Type
+		if tagType.Kind() == reflect.Ptr {
+			tagType = t.Type.Elem()
+		}
+
+		fieldName := t.Name
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply config/config_cmd.go:122]: Seems like we can't get the enum from `reflect.TypeOf`. maybe we can add all the options in pflag

File: config/config_cmd.go [hunk 8]
```diff
@@ -75,6 +103,92 @@ func redirectStdOut() (old, new *os.File) {
...
+			f := val.Field(i)
+			// In order to get value from unexported field
+			if f.Type().Kind() == reflect.Ptr {
+				f = reflect.NewAt(f.Type(), unsafe.Pointer(f.UnsafeAddr())).Elem()
+			} else {
+				f = reflect.NewAt(f.Type(), unsafe.Pointer(f.UnsafeAddr()))
+			}
+			if canPrint(f.Interface()) && sectionKeyMap[fieldType] == false {
+				sectionKeyMap[fieldType] = true
+				defer printConfigTable(f.Interface(), fieldType, true, sectionKeyMap)
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply config/config_cmd.go]: removed defer.

File: config/config_cmd.go [hunk 9]
```diff
@@ -47,12 +53,34 @@ func NewConfigCommand(accessorProvider AccessorProvider) *cobra.Command {
 		},
 	}
 
+	docsCmd := &cobra.Command{
+		Use:   "docs",
+		Short: "Generate configuration documetation in rst format",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			sections := GetRootSection().GetSections()
+			var orderedSections []string
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply config/config_cmd.go]: sorry, `sets.String` doesn't work. still get the random order when iterating the `sets.string`

[the remark this answers is not in the document]
Comment (EngHabu) [reply config/config_cmd.go]:
```golang
s := sets.NewString()
s.Insert("z")
s.Insert("a")
for _, item := range s.AsList() {
}
```

This gives you a random order?

File: config/config_cmd.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (EngHabu) [reply config/config_cmd.go:122]: Yeah, you can check the existence of `<EnumTypeName>Values()` function in the type's package to see if we've generates enumer for it..

For example, [this line](https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/plugins/array/core/state.go#L24) will generate [this file](https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/plugins/array/core/phase_enumer.go#L48)

File: config/config_cmd.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (EngHabu) [reply config/config_cmd.go]: Awesome!

File: config/config_cmd.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (EngHabu) [reply config/config_cmd.go]: I think it's almost there... some default values show up as the type name, for example, take a look at Duration here: https://flyte--1608.org.readthedocs.build/en/1608/deployment/cluster_config/flyteadmin_config.html#authorizationserver

File: config/config_cmd.go [hunk 10]
```diff
@@ -75,6 +99,155 @@ func redirectStdOut() (old, new *os.File) {
 	return
 }
 
+func printConfigTable(section Section, sectionName string, isSubsection bool, visitedSection map[string]bool, typeLink map[reflect.Type]string) {
```
Comment (EngHabu) [inline config/config_cmd.go:102]: We need to add some unit tests for this...

File: config/config_cmd.go [hunk 11]
```diff
@@ -75,6 +99,155 @@ func redirectStdOut() (old, new *os.File) {
...
+			fieldDefaultValue = fieldType
+		}
+
+		if jsonTag := t.Tag.Get("json"); len(jsonTag) > 0 && !strings.HasPrefix(jsonTag, "-") {
+			var commaIdx int
+			if commaIdx = strings.Index(jsonTag, ","); commaIdx < 0 {
+				commaIdx = len(jsonTag)
+			}
+			if jsonTag[:commaIdx] != "" {
+				fieldName = jsonTag[:commaIdx]
```
Comment (EngHabu) [inline config/config_cmd.go:139]:
```suggestion
			if strippedName := jsonTag[:commaIdx]; len(strippedName) > 0 {
				fieldName = strippedName
```

**Review by EngHabu — COMMENTED** [review]
Almost there! just responding to a couple of comments... let's add some unit tests to this function...

File: config/config_cmd.go [hunk 12]
```diff
@@ -75,6 +99,155 @@ func redirectStdOut() (old, new *os.File) {
...
+			}
+		}
+
+		if pFlag := t.Tag.Get("pflag"); len(pFlag) > 0 && !strings.HasPrefix(pFlag, "-") {
```
Comment (EngHabu) [inline config/config_cmd.go:143]: I think you can use [this class](https://github.com/flyteorg/flytestdlib/blob/master/cli/pflags/api/tag.go) to parse the tag for you... and extract the information you need... feel free to add the trimPrefix call there too..

File: config/config_cmd.go (hunk 10 above)
Comment (EngHabu) [reply config/config_cmd.go:102]: Maybe also do some refactoring to break down this function into smaller (easier to unit-test) ones?

File: config/config_cmd.go (hunk 12 above)
Comment (pingsutw) [reply config/config_cmd.go]: Can't import it. will get an error. `import cycle not allowed`

File: config/config_cmd.go (hunk 10 above)
Comment (pingsutw) [reply config/config_cmd.go]: added it

File: config/config_cmd.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (pingsutw) [reply config/config_cmd.go]: sorry, my bad. it works. somewhere I didn't use `sets.String`., so I get the random order list.

File: config/config_cmd.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (pingsutw) [reply config/config_cmd.go]: I think `<EnumTypeName>Values()` can't work here. Will find a way to fix it in separate PR.

**Comment by kumare3** [comment]
Looks awesome, thank you so much

**Review by EngHabu — APPROVED** [review]
Fantastic job
